File: en/codes/python/chapter_searching/linear_search.py
"""
File: linear_search.py
Created Time: 2022-11-26
Author: timi (xisunyy@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import ListNode, list_to_linked_list


def linear_search_array(nums: list[int], target: int) -> int:
    """Linear search (array)"""
    # Traverse the array
    for i in range(len(nums)):
        if nums[i] == target:  # Found the target element, return its index
            return i
    return -1  # Target element not found, return -1


def linear_search_linkedlist(head: ListNode, target: int) -> ListNode | None:
    """Linear search (linked list)"""
    # Traverse the linked list
    while head:
        if head.val == target:  # Found the target node, return it
            return head
        head = head.next
    return None  # Target node not found, return None


"""Driver Code"""
if __name__ == "__main__":
    target = 3

    # Perform linear search in array
    nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8]
    index: int = linear_search_array(nums, target)
    print("Index of target element 3 =", index)

    # Perform linear search in linked list
    head: ListNode = list_to_linked_list(nums)
    node: ListNode | None = linear_search_linkedlist(head, target)
    print("The corresponding node object for target node value 3 is", node)

File: ja/codes/csharp/chapter_searching/linear_search.cs
/**
 * File: linear_search.cs
 * Created Time: 2022-12-23
 * Author: haptear (haptear@hotmail.com)
 */

namespace hello_algo.chapter_searching;

public class linear_search {
    /* 線形探索（配列） */
    int LinearSearchArray(int[] nums, int target) {
        // 配列を走査
        for (int i = 0; i < nums.Length; i++) {
            // 目標要素が見つかったらそのインデックスを返す
            if (nums[i] == target)
                return i;
        }
        // 目標要素が見つからなければ -1 を返す
        return -1;
    }

    /* 線形探索（連結リスト） */
    ListNode? LinearSearchLinkedList(ListNode? head, int target) {
        // 連結リストを走査
        while (head != null) {
            // 対象ノードが見つかったら、それを返す
            if (head.val == target)
                return head;
            head = head.next;
        }
        // 対象ノードが見つからない場合は null を返す
        return null;
    }

    [Test]
    public void Test() {
        int target = 3;

        /* 配列で線形探索を行う */
        int[] nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8];
        int index = LinearSearchArray(nums, target);
        Console.WriteLine("対象要素 3 のインデックス = " + index);

        /* 連結リストで線形探索を行う */
        ListNode? head = ListNode.ArrToLinkedList(nums);
        ListNode? node = LinearSearchLinkedList(head, target);
        Console.WriteLine("目標ノード値 3 に対応するノードオブジェクトは " + node);
    }
}
